app/jobs/asset_cleanup_job.rb

```diff
@@ -2,33 +2,39 @@
 
 class AssetCleanupJob < ApplicationJob
   def perform
-    active_revision_ids = Edition
-      .where(current: true)
-      .or(Edition.where(live: true))
-      .pluck(:revision_id)
-
-    active_file_attachment_asset_ids = FileAttachment::Revision
-      .joins(:revisions, blob_revision: [:asset])
-      .where("revisions.id" => active_revision_ids)
-      .pluck("file_attachment_assets.id")
+    clean_image_assets
+    clean_file_attachment_assets
+  end
 
-    active_image_asset_ids = Image::Revision
-      .joins(:revisions, blob_revision: [:assets])
-      .where("revisions.id" => active_revision_ids)
-      .pluck("image_assets.id")
+private
 
-    FileAttachment::Asset
-      .where.not(id: active_file_attachment_asset_ids)
-      .where.not(state: "absent")
-      .each { |asset| delete_asset(asset) }
+  def clean_image_assets
+    image_asset_used =
+      Edition.where(current: true)
+             .or(Edition.where(live: true))
+             .joins(revision: { image_revisions: :blob_revision })
+             .where("image_assets.blob_revision_id = image_blob_revisions.id")
+             .arel
```

Comment:
I don't really understand this line and the one below, could you explain if thats okay?

Reply:
No problem. [Arel](https://www.rubydoc.info/github/rails/arel) is the underlying SQL generation library that Rails uses.

`.exists` is a method that will turn a SELECT query into one embedded in a EXISTS condition. For example: 

```
irb(main):010:0> Edition.where(id: 404).arel.to_sql
=> "SELECT \"editions\".* FROM \"editions\" WHERE \"editions\".\"id\" = $1"
irb(main):011:0> Edition.where(id: 404).arel.exists.to_sql
=> "EXISTS (SELECT \"editions\".* FROM \"editions\" WHERE \"editions\".\"id\" = $1)"
```

You used to be able to run `exists` directly on an ActiveRecord relation but this is now deprecated:

```
irb(main):012:0> Edition.where(id: 404).exists.to_sql
DEPRECATION WARNING: Delegating exists to arel is deprecated and will be removed in Rails 6.0. (called from irb_binding at (irb):12)
=> "EXISTS (SELECT \"editions\".* FROM \"editions\" WHERE \"editions\".\"id\" = $1)"
```

So anyway this gets us around to what this query is doing. We're built up a sub query that checks whether an image_asset is associated with a live or current edition. We're then negating that sub query to check whether there are any image_assets that don't have that association, so that we know it's safe to delete them.

Reply:
Thanks for the explanation, it has been really helpful! I wonder why they're deprecating the exists call on a relation as it seems weird to have to have to call to arel before using it.